api/v2/api.go

```diff
@@ -285,11 +278,24 @@ type TLSConfig struct {
 }
 
 // ListenAndServe spins up the API server
-func (api *API) ListenAndServe(addr string, tls *TLSConfig) error {
-	if tls != nil {
-		return api.r.RunTLS(addr, tls.CertFile, tls.KeyFile)
+func (api *API) ListenAndServe(ctx context.Context, addr string, tls *TLSConfig) error {
+	errChan := make(chan error, 1)
+	go func() {
+		if tls != nil {
+			errChan <- api.r.RunTLS(addr, tls.CertFile, tls.KeyFile)
+			return
+		}
+		errChan <- api.r.Run(addr)
+		return
+	}()
+	for {
+		select {
+		case err := <-errChan:
+			return err
+		case <-ctx.Done():
+			return nil
```

Comment:
run this select in a goroutine instead, do `api.r.Close` when ctx finishes

Reply:
actually i might be wrong on that... at a glance it doesn't look like gin offers a clean `Close` 😱

Reply:
:cry:  As mentioned over slack, in another PR we'll do somethign similar to https://github.com/gin-gonic/gin#graceful-restart-or-stop so we can have non-janky+graceful closures.